### ISSUE TITLE
MLX Phase 2A: Model Adapter And MoE Weight Layout Contract

### ISSUE THREAD

**Issue by egesabanci** [egesabanci/reap-mlx#4]
## Objective

Create the MLX model-adapter and weight-layout contract that discovers MoE layers, represents arbitrary MoE routing/expert layouts, extracts architecture-specific config, builds attention masks, and provides the extension points needed by the future MLX observer and pruner.

## Background

MLX does not support the same hook-oriented execution model as PyTorch, and MLX-LM architectures and user-provided MoE weights can expose different runtime module attributes and tensor layouts. A small adapter layer is needed so observer and pruning code operate through explicit layout/routing contracts instead of ad hoc reflection or hardcoded model names.

This should start with Qwen3-MoE but keep extension points explicit for Mixtral, DeepSeek, GLM, and ERNIE.

## Scope

Add model utility/adapters only. No full observer replay and no pruning mutation in this issue.

Target files may include:

- `src/reap/backends/mlx/model_util.py`
- `src/reap/backends/mlx/model_adapters.py` if a separate adapter module is cleaner
- focused tests, e.g. `tests/test_mlx_model_util.py`

## Implementation Steps

1. Implement `_identify_moe_layers(model)` or an adapter method equivalent.
2. Support common MLX-LM model layouts where layers live under `model.model.layers` or equivalent adapter-provided paths.
3. Detect MoE layers by presence of routed/stacked MoE structures such as `switch_mlp`, while allowing architecture-specific overrides.
4. Add helpers to get the MoE module for a layer.
5. Add helpers to get dense MLP fallback modules for non-MoE layers.
6. Add helpers to extract per-layer config:
   - number of experts
   - top-k / experts per token
   - normalization behavior
   - architecture name or adapter type
7. Add helpers to update config dicts after pruning without assuming `model.config` exists.
8. Add attention-mask helper for explicit layer replay. For MLX-LM Qwen3-style layers, prefer the same causal mask path used by MLX-LM normal forward.
9. Add shared-expert detection helper that checks both `shared_expert` and `shared_experts`.
10. Keep imports MLX-safe and localize any `mlx` / `mlx_lm` imports so the package skeleton remains safe.

## Acceptance Criteria

- [ ] MoE layer discovery returns correct layer indices for a tiny/synthetic Qwen3-style model.
- [ ] Dense-only layers return no MoE layer indices.
- [ ] Config extraction prefers live module attributes over stale config dict values.
- [ ] Top-k config lookup works with both `num_experts_per_tok` and `top_k` conventions.
- [ ] Shared expert detection handles both singular and plural attribute names.
- [ ] Config update helpers operate on the dict returned by `mlx_lm.load(..., return_config=True)`.
- [ ] No code assumes `model.config` exists.
- [ ] Importing the module does not import `torch` or `vllm`.

## Test Plan

Run:

```bash
pytest tests/test_mlx_no_torch_import.py tests/test_mlx_model_util.py -v
```

Suggested tests:

- tiny model with all layers MoE
- tiny model with partial MoE layers
- dense-only model
- config extraction before and after simulated top-k clamp
- shared expert singular/plural detection

## Non-Goals

- Do not implement the full observer.
- Do not implement router adapters beyond what already exists.
- Do not implement pruning tensor slicing.
- Do not download real models in unit tests.
- Do not refactor existing Torch `MODEL_ATTRS`.

## References

- `implementation-plan-draft.md`, P2-001
- `docs/reap-mlx-implementation-plan-gpt-v0.2.md`, Layer Replay and Config Update

---

## Portfolio Objective Alignment - 2026-05-31

This issue is part of the Apple Silicon MLX REAP backend effort. The production/official REAP experimentation workflow remains the original PyTorch/CUDA implementation; this MLX work must not destabilize or replace that path.

The MLX backend goal is adapter-driven support for MoE weights, not a one-model port. Qwen3-specific work is bootstrap/reference-adapter work. The implementation should preserve REAP pruning semantics and observer-data schema compatibility while adapting to MLX constraints such as lazy execution, no PyTorch hooks, MLX-LM module layouts, and save/reload behavior.

For this issue, the adapter layer is the main mechanism for supporting arbitrary MoE weights. It must define clear extension points for routing, layer discovery, expert execution, pruning fields, and config updates instead of hardcoding one model family.
